effekt/shared/src/main/scala/effekt/core/Transformer.scala

```diff
@@ -185,8 +191,13 @@ object Transformer extends Phase[Typechecked, CoreTransformed] {
         val bparams = bps map transform
         Def(f.symbol, BlockLit(tparams, cparams, vparams, bparams, transform(body)), transform(rest))
 
-      case v @ source.ValDef(id, _, binding) if pureOrIO(binding) =>
-        Let(v.symbol, Run(transform(binding)), transform(rest))
+      case v @ source.ValDef(id, tpe, binding) if pureOrIO(binding) =>
+        val transformed = Run(transform(binding))
+        val transformedTpe = v.symbol.tpe match {
+          case Some(tpe) => transform(tpe)
+          case None => transformed.tpe
+        }
```

Comment:
This pattern shows up quite often, how can we simplify?

Reply:
Something like `foo(v.symbol.tpe, transformed.tpe)` might be possible, equivalent to a
`v.symbol.tpe.map(transform).getOrElse(transformed.tpe)`.
Though I'm not sure it's worth it.

I am unsure if we can abstract over more easily since the type of `binding`/`transformed` is different in most cases (`Stmt` for `val`, `Expr` for `let`). - This also means that the call to transform (on the binding) is a different one.